cmd/kubeadm/app/master/selfhosted_volumes.go

```diff
@@ -0,0 +1,294 @@
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func schedulerProjectedVolume() *v1.Volume {
+	return &v1.Volume{
+		Name: "k8s",
+		VolumeSource: v1.VolumeSource{
+			Projected: &v1.ProjectedVolumeSource{
+				Sources: []v1.VolumeProjection{
+					{
+						Secret: &v1.SecretProjection{
+							LocalObjectReference: v1.LocalObjectReference{
+								Name: kubeadmconstants.SchedulerKubeConfigFileName,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func controllerManagerProjectedVolume() *v1.Volume {
+	return &v1.Volume{
+		Name: "k8s",
+		VolumeSource: v1.VolumeSource{
+			Projected: &v1.ProjectedVolumeSource{
+				Sources: []v1.VolumeProjection{
+					{
+						Secret: &v1.SecretProjection{
+							LocalObjectReference: v1.LocalObjectReference{
+								Name: kubeadmconstants.ControllerManagerKubeConfigFileName,
+							},
+						},
+					},
+					{
+						Secret: &v1.SecretProjection{
+							LocalObjectReference: v1.LocalObjectReference{
+								Name: kubeadmconstants.CACertAndKeyBaseName,
+							},
+							Items: []v1.KeyToPath{
+								{
+									Key:  "tls.crt",
+									Path: path.Join("pki", kubeadmconstants.CACertName),
+								},
+								{
+									Key:  "tls.key",
+									Path: path.Join("pki", kubeadmconstants.CAKeyName),
+								},
+							},
+						},
+					},
+					{
+						Secret: &v1.SecretProjection{
+							LocalObjectReference: v1.LocalObjectReference{
+								Name: kubeadmconstants.ServiceAccountKeyBaseName,
+							},
+							Items: []v1.KeyToPath{
+								{
+									Key:  "tls.key",
+									Path: path.Join("pki", kubeadmconstants.ServiceAccountPrivateKeyName),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func createTLSSecrets(client *clientset.Clientset) error {
+	tlsPairs := []string{
+		kubeadmconstants.CACertAndKeyBaseName,
+		kubeadmconstants.CACertName,
+		kubeadmconstants.CAKeyName,
+		kubeadmconstants.APIServerCertAndKeyBaseName,
+		kubeadmconstants.APIServerCertName,
+		kubeadmconstants.APIServerKeyName,
+		kubeadmconstants.APIServerKubeletClientCertAndKeyBaseName,
+		kubeadmconstants.APIServerKubeletClientCertName,
+		kubeadmconstants.APIServerKubeletClientKeyName,
+		kubeadmconstants.ServiceAccountKeyBaseName,
+		kubeadmconstants.ServiceAccountPublicKeyName,
+		kubeadmconstants.ServiceAccountPrivateKeyName,
+		kubeadmconstants.FrontProxyCACertAndKeyBaseName,
+		kubeadmconstants.FrontProxyCACertName,
+		kubeadmconstants.FrontProxyCAKeyName,
+	}
+	for i := 0; i < len(tlsPairs); i = i + 3 {
+		fmt.Printf("[self-hosted] Creating TLS secret %q with %s and %s\n", tlsPairs[i], tlsPairs[i+1], tlsPairs[i+2])
+		secret, err := createTLSSecretFromFiles(
+			tlsPairs[i],
+			path.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, "pki", tlsPairs[i+1]),
+			path.Join(kubeadmapi.GlobalEnvParams.KubernetesDir, "pki", tlsPairs[i+2]),
+		)
+		if err != nil {
+			return err
+		}
+
+		if _, err := client.Secrets(metav1.NamespaceSystem).Create(secret); err != nil {
```

Comment:
The pattern when invoking the client should be `client.APIGroupVersion().Resource().Verb()`.
This should be `client.CoreV1().Secrets()...`, usage of `client.Secrets()` is a deprecated pattern IIRC